pkg/endpoint/endpoint.go

```diff
@@ -1829,7 +1833,12 @@ func (e *Endpoint) setPolicyRevision(rev uint64) {
 			ps.done(now)
 			delete(e.policyRevisionSignals, ps)
 		default:
-			if rev >= ps.wantedRev {
+			// The policy revision of an endpoint can be bumped
+			// before the initial build has happened if an endpoint
+			// is not affected by a policy update. Require an
+			// initial build to have happened to assume that a
+			// policy revision is implemented.
+			if rev >= ps.wantedRev && e.initialBuildComplete {
```

Comment:
Is there a meaningful distinction by adding the check here rather than exiting the function in the check at the top?

Seems like line 1825 is actually setting the policy revision in this case, even though this change is no longer notifying any goroutine that's listening for the signal that this version is implemented.

Reply:
> Is there a meaningful distinction by adding the check here rather than exiting the function in the check at the top?

Initially, I thought it would be correct to still set the policy revision but thinking about again, it seems more accurate to only set it once the initial build has been done.

I was also worried about not cleaning up canceled contexts anymore but we are guaranteed to have another build still running which will ultimately clean up.

I'll change it. Good find.

Reply:
Looking into this further, I have validated that the endpoint event queue ensures that a policy update cannot get in front of an already queued regeneration. This resolves the concern that by skipping the update of the revision, a sequence such as `setPolicyRevision(1)`, `setPolicyRevision(3)`, `setPolicyRevision(2)` would miss revision 3.

However, this is not guaranteed initially as the endpoint is being added to the endpointmanager before the initial regeneration is being triggered. It can thus be found by `GetPolicyEndpoints()` and the policy layer can queue an event before the initial regeneration is queued onto the endpoint events queue.

This means that we can have a sequence like this. It seems that it is fine though:

0. policyRev = 0
1. CNI ADD
2. endpointmanager.AddEndpoint()
3. policy update which does not affect endpoint being created
4. policyRev = 1
5. endpoint.PolicyRevisionBumpEven(1)
6. endpoint.RegenerateAfterCreation() // triggered by CNI add
7. setPolicyRevision(1) // via endpoint.PolicyRevisionBumpEven(1), skipped due to missing initial build
8. setPolicyRevision(1) // via  endpoint.RegenerateAfterCreation() 

It seems guaranteed that the policy revision used by `endpoint.RegenerateAfterCreation()` cannot be lesser than any policy revision bump already queued.

As a consequence, it seems like a good idea for `GetPolicyEndpoints()` to *not* return endpoints which have not been built yet so we never attempt to skip policy revision builds until the initial build has been done.